azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -223,73 +223,141 @@ def _handle_pipeline_event(self, event):
-    requests until that state is achieved.
+    it needs to be connected.
     """
 
     def __init__(self):
         super(EnsureConnectionStage, self).__init__()
         self.connected = False
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _execute_op(self, op):
+        # Any operation that requires a connection can trigger a connection if
+        # we're not connected.
+        if op.needs_connection and not self.connected:
+            self._do_connect(op)
+
+        # Finally, if this stage doesn't need to do anything else with this operation,
+        # it just passes it down.
+        else:
+            operation_flow.pass_op_to_next_stage(self, op)
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _do_connect(self, op):
+        """
+        Start connecting the transport in response to some operation
+        """
+        # function that gets called after we're connected.
+        @pipeline_thread.runs_on_pipeline_thread
+        def on_connected(op_connect):
+            logger.info(
+                "{}({}): connection is complete: {}".format(self.name, op.name, op_connect.error)
+            )
+            if op_connect.error:
+                logger.info(
+                    "{}({}): completing with failure because of connection failure".format(
+                        self.name, op.name
+                    )
+                )
+                op.error = op_connect.error
+                operation_flow.complete_op(stage=self, op=op)
+            else:
+                logger.info(
+                    "{}({}): connection is complete.  Continuing with op".format(self.name, op.name)
+                )
+                operation_flow.pass_op_to_next_stage(stage=self, op=op)
+
+        # call down to the next stage to connect.
+        logger.info("{}({}): calling down with Connect operation".format(self.name, op.name))
+        operation_flow.pass_op_to_next_stage(
+            self, pipeline_ops_base.ConnectOperation(callback=on_connected)
+        )
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def on_connected(self):
+        self.connected = True
+        PipelineStage.on_connected(self)
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def on_disconnected(self):
+        self.connected = False
+        PipelineStage.on_disconnected(self)
+
+
+class SerializeConnectOpsStage(PipelineStage):
+    """
+    This stage is responsible for serializing connect, disconnect, and reconnect ops on
+    the pipeline, such that only a single one of these ops can go past this stage at a
+    time.  This way, we don't have to worry about cases like "what happens if we try to
+    disconnect if we're in the middle of reconnecting."  This stage will wait for the
+    reconnect to complete before letting the disconnect past.
+    """
+
+    def __init__(self):
+        super(SerializeConnectOpsStage, self).__init__()
         self.queue = queue.Queue()
         self.blocked = False
+        self.connected = False
```

Comment:
Should this maybe be maintained on the Root rather than on multiple individual stages?

Reply:
Yes, I was debating about that.  Good idea.

---
In reply to: [311702803](https://github.com/Azure/azure-iot-sdk-python-preview/pull/160#discussion_r311702803) [](ancestors = 311702803)